pkg/kubectl/cmd/apply.go

```diff
@@ -585,33 +599,55 @@ func (p *patcher) patchSimple(obj runtime.Object, modified []byte, source, names
 		return nil, nil, cmdutil.AddSourceToErr(fmt.Sprintf("retrieving original configuration from:\n%v\nfor:", obj), source, err)
 	}
 
-	// Create the versioned struct from the type defined in the restmapping
-	// (which is the API version we'll be submitting the patch to)
-	versionedObject, err := scheme.Scheme.New(p.mapping.GroupVersionKind)
 	var patchType types.PatchType
 	var patch []byte
+	var lookupPatchMeta strategicpatch.LookupPatchMeta
+	var schema oapi.Schema
 	createPatchErrFormat := "creating patch with:\noriginal:\n%s\nmodified:\n%s\ncurrent:\n%s\nfor:"
-	switch {
-	case runtime.IsNotRegisteredError(err):
-		// fall back to generic JSON merge patch
-		patchType = types.MergePatchType
-		preconditions := []mergepatch.PreconditionFunc{mergepatch.RequireKeyUnchanged("apiVersion"),
-			mergepatch.RequireKeyUnchanged("kind"), mergepatch.RequireMetadataKeyUnchanged("name")}
-		patch, err = jsonmergepatch.CreateThreeWayJSONMergePatch(original, modified, current, preconditions...)
-		if err != nil {
-			if mergepatch.IsPreconditionFailed(err) {
-				return nil, nil, fmt.Errorf("%s", "At least one of apiVersion, kind and name was changed")
+	// Try to use openapi first if the openapi spec is available and can successfully calculate the patch.
+	// Otherwise, fall back to baked-in types.
+	if p.openapiSchema != nil {
+		if schema = p.openapiSchema.LookupResource(p.mapping.GroupVersionKind); schema != nil {
```

Comment:
is there any indication in the discovery doc that the resource supports strategic merge patch? I would not assume the presence of an openapi schema to mean SMP is supported.

Reply:
I guess this question is related to the discussion https://github.com/kubernetes/kubernetes/pull/56606#issuecomment-348202176.

> is there any indication in the discovery doc that the resource supports strategic merge patch?

I don't think there is an explicit indicator.
If there are patch metadata (e.g. PatchStrategy, MergeKey) in the discovery doc, then this type supports SMP. But if there is not patch metadata, the type may still support SMP and it uses the default behavior(replacing lists).

> I would not assume the mere presence of an openapi schema to mean SMP is supported.

It's not clear what does `SMP is supported` mean.
If a type and its schema appear in the openapi schema with no patch metadada, SMP will process it with the default approach (replacing a list). It should not do anything wrong.

Did I miss anything?

Reply:
Yes, it assumes every API server that publishes openapi schemas is able to accept strategic merge patches. Previously, the client only assumed that compiled-in known core kubernetes types were served by a server supporting SMP.

Reply:
Let me summarize it for SMP

Before this PR:
Only compiled-in types (native k8s api resources) are supported.
No support for user-defined api resource no support for CRD.

After this PR:
There are 2 approaches:
Using baked-in types: only support native k8s api resources
Using openapi schema: support native k8s api resources and user-defined api resource. But no support for CRD.

Correct me if anything is wrong.

Reply:
> Using openapi schema: support native k8s api resources and user-defined api resource

It assumes user-defined api resources support strategic merge patch

Reply:
> It assumes user-defined api resources support strategic merge patch

Yes. Even though there may be no patch metadata defined in user-defined api resources, SMP will simply use default behavior which is similar to JSON merge patch as long as it has openapi schema.

I don't understand in what scenario this may be a problem. Can you clarify your concern?

Reply:
> I don't understand in what scenario this may be a problem. Can you clarify your concern?

This is not checking the "consumes" portion of the published openapi spec to see if strategic merge patches are accepted. If the patch operation on the resource does not indicate it consumes "application/strategic-merge-patch+json", we shouldn't assume we can send it a strategic merge patch.

```
    "patch": {
     "description": "partially update the specified ConfigMap",
     "consumes": [
      "application/json-patch+json",
      "application/merge-patch+json",
      "application/strategic-merge-patch+json"
     ],
```

Reply:
Got your point now. 

Do you think that checking server accepted types logic is necessary for 1.9 release? If so, I will start to work on that today.

Reply:
If we want to start sending SMP requests based on openapi schemas, yes

Reply:
Did this PR change whether kubectl sends a SMP vs a JP based on the presence of openapi data?

Reply:
Yes. Previously, non-compiled types would send JP. Now if there is an openapi schema, it sends SMP unless it errors building the patch. It should also check for consumes SMP type patch

Reply:
There is a potential simpler way to make this work:
- construct the SMP from openapi schema even for a non-compiled type
- send it to the server with MIME type: SMP
- server reject it and return some HTTP error code (maybe 406)
- retry with the JMP

Is this way acceptable?

Reply:
That doesn't sound simpler than using the information about the accepted mime types that is in the discovery doc we already have here. That's now how we do other negotiation (e.g. "send various versions of an API object until we find one the server doesn't return an http error for")

Reply:
I agree with a jordan here.  Mime type negotiation exists for a reason.  We shouldn’t invent a new way to do something that already exists.

Reply:
I agree the best practice should be MIME type negotiation.

But to do MIME type negotiation, we need to parse the `endpoints` part of openapi schema.
There is no facility to support parsing the `endpoints` part of openapi schema, we only have the code to parse the `models` part. 
So we will need to introduce quite some new code to support that. So this way will require much more efforts than the approach I mentioned above. 
IMO it may be risky to introduce a lot of new code at this point.

Reply:
> But to do MIME type negotiation, we need to parse the endpoints part of openapi schema.

If we want to use openapi docs dynamically to form patch requests, I think that's what we need to do.

> So this way will require much more efforts than the approach I mentioned above.
> IMO it may be risky to introduce a lot of new code at this point.

I agree it is more effort, but I have a hard time accepting the "if at first you don't succeed" negotiation approach when we've already been given exactly the information we need to know what patch types are accepted

> and is risky at this point

I also agree with this. Which is better at this point for this release?
* attempting multiple patch types falling back to merge patch if SMP is rejected with 406?
* defaulting this feature off
* disabling this feature

Whatever we choose for the moment, we should work to make use of the accepted content in the next release.

Reply:
Can we: send SMP if the type is registered, JMP otherwise? AFAIU That wouldn't really change the existing behavior (before that pull-request), yet exercise that code and already have some benefits?

Reply:
This is a good example of why we need libraries that combine the discovery and openapi information for each resource type.  We need and abstraction that hides the specific source of metadata for the resources.

@liggitt What do you recommend for 1.9?  Disable or try to get a fix in?

Reply:
I'm leaning to what @apelisse suggested in https://github.com/kubernetes/kubernetes/pull/51321#discussion_r154271945.
It should be a trivial change with fewer than 50 lines of code change.

Reply:
hi @liggitt @mengqiy about the MIME type negotiation part is that means we check the "consumes" portion of a related resource in discovery doc, if it support, then use SMP, if not use JMP.  if so what we need maybe in 1.10 is to add the function as @mengqiy memtioned

>But to do MIME type negotiation, we need to parse the endpoints part of openapi schema.
There is no facility to support parsing the endpoints part of openapi schema, we only have the code to parse the models part. 

so all we have to do is make `NewOpenAPIData` also return `paths` and add some lookup/indication function ?  @mengqiy @apelisse  I would like to take a look of this openapi helper function if you haven't yet.

>This is a good example of why we need libraries that combine the discovery and openapi information for each resource type. We need and abstraction that hides the specific source of metadata for the resources.

This also sounds interesting, but I haven't got the idea of "combine" @pwittrock could you explain more ? thanks